### ISSUE TITLE
Add option to disallow trailing data

### ISSUE THREAD

**Issue by simolus3** [simolus3/tar#11]
Tar files end with two zero blocks where a header would be expected. When we encounter this, the tar reader automatically closes itself and cancels the underlying stream subscription.

As we found out in https://github.com/dart-lang/pub/pull/2935, this can cause issues for users expecting the reader to always read the stream fully. There could be an option to asset that no further data is received at the end of a tar stream. When that option is enabled, we could wait for further data or the end of the stream and throw an error in the first case.

**Comment by simolus3** [comment]
@isoos In the pub server, you're only reading tar streams from memory or a file, right? The reader used to always cancel the underlying stream subscription when encountering an end-of-archive marker. Cancelling response streams had unintended side-effects in the http client, I'm worried something similar might happen when cancelling the request stream in the server. If you fully buffered the stream before there's no concern here.

I've released version `0.3.1` of the `tar` package which adds a parameter to not cancel the subscription but throw when there's unexpected data after a marker indicating the end of an archive.

**Comment by isoos** [comment]
Yeah, pub-dev is reading from file (in prod) or from memory (in tests). Thanks for the update, I'll add a test for that. If I understand this properly, this will throw if I append an extra byte to any valid package archive. Am I right with that?

**Comment by simolus3** [comment]
> If I understand this properly, this will throw if I append an extra byte to any valid package archive. Am I right with that?

Yes, exactly. Note that throwing will _still_ cancel the stream subscription: 
https://github.com/simolus3/tar/blob/fbe1b9465ace95a525023260c5bc57722f659e6e/test/reader_test.dart#L134-L144

If you're only reading from files or from memory, it's not necessary to set the `disallowTrailingData` flag, since there's no harm when cancelling a stream (the file descriptor will be closed either way). For http clients it resets the TCP connection which was a real issue.
I also haven't checked if there are any existing packages on pub.dev that might break with `disallowTrailingData` since this is kind of a pedantic requirement for tar files. @jonasfj Suggested that such check could make sense for new packages though, if that's the only place where tar files are decoded.

**Comment by isoos** [comment]
@simolus3 could you please help me out what the problem may be? Is it that we have a tar archive with bad content, or maybe some edge case in the code?

```dart
import 'dart:io';

import 'package:http/http.dart';
import 'package:tar/tar.dart';

Future<void> main() async {
  final rs = await get(Uri.parse(
      'https://pub.dartlang.org/packages/pana/versions/0.12.19.tar.gz'));
  final uncompressed = gzip.decode(rs.bodyBytes);
  // works
  print('default...');
  await _read(TarReader(Stream.fromIterable([uncompressed])));

  // fails
  print('disallowTrailingData...');
  await _read(TarReader(
    Stream.fromIterable([uncompressed]),
    disallowTrailingData: true,
  ));
}

Future<void> _read(TarReader reader) async {
  while (await reader.moveNext()) {
    print('- ${reader.current.name}');
  }
}
```

**Comment by simolus3** [comment]
It looks like the tar file has a lot more than two empty blocks at the end

```
cat reference/pub/pana-0.12.19.tar.gz | gzip -d | tail -c 6500 | hexdump -C

00000000  20 20 20 20 65 78 70 65  63 74 28 6e 6f 6e 41 73  |    expect(nonAs|
00000010  63 69 69 52 75 6e 65 52  61 74 69 6f 28 27 20 20  |ciiRuneRatio('  |
00000020  5c 74 5c 6e 5c 72 27 29  2c 20 30 2e 30 29 3b 0a  |\t\n\r'), 0.0);.|
00000030  20 20 20 20 7d 29 3b 0a  0a 20 20 20 20 74 65 73  |    });..    tes|
00000040  74 28 27 61 73 63 69 69  20 74 65 78 74 27 2c 20  |t('ascii text', |
00000050  28 29 20 7b 0a 20 20 20  20 20 20 65 78 70 65 63  |() {.      expec|
00000060  74 28 6e 6f 6e 41 73 63  69 69 52 75 6e 65 52 61  |t(nonAsciiRuneRa|
00000070  74 69 6f 28 27 61 27 29  2c 20 30 2e 30 29 3b 0a  |tio('a'), 0.0);.|
00000080  20 20 20 20 20 20 65 78  70 65 63 74 28 6e 6f 6e  |      expect(non|
00000090  41 73 63 69 69 52 75 6e  65 52 61 74 69 6f 28 27  |AsciiRuneRatio('|
000000a0  61 20 62 20 63 27 29 2c  20 30 2e 30 29 3b 0a 20  |a b c'), 0.0);. |
000000b0  20 20 20 7d 29 3b 0a 0a  20 20 20 20 74 65 73 74  |   });..    test|
000000c0  28 27 6e 6f 6e 2d 61 73  63 69 69 20 74 65 78 74  |('non-ascii text|
000000d0  27 2c 20 28 29 20 7b 0a  20 20 20 20 20 20 65 78  |', () {.      ex|
000000e0  70 65 63 74 28 6e 6f 6e  41 73 63 69 69 52 75 6e  |pect(nonAsciiRun|
000000f0  65 52 61 74 69 6f 28 27  e5 b0 81 e8 a3 85 68 74  |eRatio('......ht|
00000100  74 70 e4 b8 9a e5 8a a1  e6 8e a5 e5 8f a3 27 29  |tp............')|
00000110  2c 20 30 2e 36 29 3b 0a  20 20 20 20 7d 29 3b 0a  |, 0.6);.    });.|
00000120  20 20 7d 29 3b 0a 7d 0a  00 00 00 00 00 00 00 00  |  });.}.........|
00000130  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00001960  00 00 00 00                                       |....|
```

I think we can allow more than two zero blocks at the end, I'll change the implementation.

**Comment by simolus3** [comment]
Thanks again for the repro. I've fixed this in 1aa9e4d4ae7aa6f3e6640db17018f75f8d87fbea and subsequently published version 0.3.2 with the fix.

**Comment by isoos** [comment]
@simolus3 Fantastic, thank you. Filed https://github.com/dart-lang/pub-dev/pull/4719 as the first step to check the extra bytes in pub.dev.
